colexecjoin: optimize merge/cross joins

yuzefovich opened the pull request:
**colexecjoin: make cross/merge join streaming with regards to left input**

This commit refactors the cross and merge join to be streaming with
regards to the left input. Previously, we were using two spilling queues
to consume both inputs first before proceeding to building the cross
product (in case of the merge join this is needed when building from the
buffered group).

That approach is suboptimal because buffering only one side is
sufficient, so this commit switches the cross join builder to operate in
a streaming fashion with regards to the left input. This is done by
building all result rows that correspond to the current left batch
before proceeding to the next left batch and allows us to significantly
reduce amount of copying and, thus, improving the performance.

Fixes: #67816.

Release note: None

**colexecjoin: improve probing in the merge joiner with nulls**

For non set-operation joins whenever we have nulls in both columns we
can advance both pointers since neither of the rows will have a match.
This commit takes advantage of this observation as well as refactors
(hopefully making it cleaner) the probing mechanism a bit.

Release note: None

**colexecjoin: avoid buffering tuples from the right in merge joiner**

Depending on the join type, we don't need to fully buffer the tuples
from the right input in order to produce the output. Namely, for
set-operation joins we only need to know the number of right tuples
whereas for LEFT SEMI and RIGHT ANTI we know exactly the behavior of the
builder for the buffered group.

Release note: None

**colexecjoin: remove a copy when buffering the right group**

Previously, before enqueueing the tuples from the right buffered group
into the spiling queue we would perform a deep-copy. This is an overkill
because the spilling queue itself performs the deep copy. This commit
refactors the enqueueing code to modify the right batch directly to
include only the tuples from the group.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/68921)
<!-- Reviewable:end -->

yuzefovich committed:
colexecjoin: improve probing in the merge joiner with nulls

For non set-operation joins whenever we have nulls in both columns we
can advance both pointers since neither of the rows will have a match.
This commit takes advantage of this observation as well as refactors
(hopefully making it cleaner) the probing mechanism a bit.

Release note: None

yuzefovich committed:
colexecjoin: avoid buffering tuples from the right in merge joiner

Depending on the join type, we don't need to fully buffer the tuples
from the right input in order to produce the output. Namely, for
set-operation joins we only need to know the number of right tuples
whereas for LEFT SEMI and RIGHT ANTI we know exactly the behavior of the
builder for the buffered group.

Release note: None

yuzefovich committed:
colexecjoin: remove a copy when buffering the right group

Previously, before enqueueing the tuples from the right buffered group
into the spiling queue we would perform a deep-copy. This is an overkill
because the spilling queue itself performs the deep copy. This commit
refactors the enqueueing code to modify the right batch directly to
include only the tuples from the group.

Release note: None

yuzefovich commented:
The benchmarks are [here](https://gist.github.com/yuzefovich/b2e99fcca6f7582c0edf2909734fbc87).

yuzefovich commented:
@michae2 tagging you as the main reviewer, but if anyone else is interested in diving into the vectorized merge joiner - please let me know :) (I think only Jordan had some context earlier and probably nobody but me still has it :/ )

yuzefovich commented:
Rebased on top of master. PTAL.

yuzefovich commented:
cc @rytaft @cucaroach

yuzefovich commented:
Rebased on top of master just in case.

michae2 reviewed (changes requested):
Still reading, just noticed one thing so far.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-Mm5Pb3I6NJZ2f02Rn9X:b-o0o31y)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 45 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-Mm4xJRl7VaLpUVE9pnS:-Mm4xJRl7VaLpUVE9pnT:b44mhb3) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fec4696339a4b33b94885f7c4d5fccd6062f35/pkg/sql/colexec/colexecjoin/crossjoiner.go#L45)):*
> "set-op cross joins are invalid"

Are you sure this is true? :smiling_imp: For example:

```sql
CREATE TABLE t ();
CREATE TABLE u ();
INSERT INTO t DEFAULT VALUES;
INSERT INTO t DEFAULT VALUES;
INSERT INTO u DEFAULT VALUES;
SELECT * FROM t INTERSECT ALL SELECT * FROM u;
```

Looks like it uses cross join:
```
demo@127.0.0.1:26257/defaultdb> EXPLAIN SELECT * FROM t INTERSECT ALL SELECT * FROM u;
                                        info
------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • intersect all
  │
  ├── • scan
  │     estimated row count: 2 (100% of the table; stats collected 6 seconds ago)
  │     table: t@t_pkey
  │     spans: FULL SCAN
  │
  └── • scan
        missing stats
        table: u@u_pkey
        spans: FULL SCAN
(14 rows)


Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> EXPLAIN (VEC) SELECT * FROM t INTERSECT ALL SELECT * FROM u;
               info
----------------------------------
  │
  └ Node 1
    └ *colexecjoin.crossJoiner
      ├ *colfetcher.ColBatchScan
      └ *colfetcher.ColBatchScan
(5 rows)
```


<!-- Sent from Reviewable.io -->

yuzefovich committed:
colexecjoin: make cross/merge join streaming with regards to left input

This commit refactors the cross and merge join to be streaming with
regards to the left input. Previously, we were using two spilling queues
to consume both inputs first before proceeding to building the cross
product (in case of the merge join this is needed when building from the
buffered group).

That approach is suboptimal because buffering only one side is
sufficient, so this commit switches the cross join builder to operate in
a streaming fashion with regards to the left input. This is done by
building all result rows that correspond to the current left batch
before proceeding to the next left batch and allows us to significantly
reduce amount of copying and, thus, improving the performance.

Release note: None

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 45 at r14](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-Mm4xJRl7VaLpUVE9pnS:-Mm5lUwg0E-Cl60rBReR:bca65bk) ([raw file](https://github.com/cockroachdb/cockroach/blob/04fec4696339a4b33b94885f7c4d5fccd6062f35/pkg/sql/colexec/colexecjoin/crossjoiner.go#L45)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> "set-op cross joins are invalid"

Are you sure this is true? :smiling_imp: For example:

```sql
CREATE TABLE t ();
CREATE TABLE u ();
INSERT INTO t DEFAULT VALUES;
INSERT INTO t DEFAULT VALUES;
INSERT INTO u DEFAULT VALUES;
SELECT * FROM t INTERSECT ALL SELECT * FROM u;
```

Looks like it uses cross join:
```
demo@127.0.0.1:26257/defaultdb> EXPLAIN SELECT * FROM t INTERSECT ALL SELECT * FROM u;
                                        info
------------------------------------------------------------------------------------
  distribution: local
  vectorized: true

  • intersect all
  │
  ├── • scan
  │     estimated row count: 2 (100% of the table; stats collected 6 seconds ago)
  │     table: t@t_pkey
  │     spans: FULL SCAN
  │
  └── • scan
        missing stats
        table: u@u_pkey
        spans: FULL SCAN
(14 rows)


Time: 1ms total (execution 1ms / network 0ms)

demo@127.0.0.1:26257/defaultdb> EXPLAIN (VEC) SELECT * FROM t INTERSECT ALL SELECT * FROM u;
               info
----------------------------------
  │
  └ Node 1
    └ *colexecjoin.crossJoiner
      ├ *colfetcher.ColBatchScan
      └ *colfetcher.ColBatchScan
(5 rows)
```
</blockquote></details>

Indeed, nice catch! Fixed.

It's interesting that the set-op cross join occurs only when the tables don't have any visible columns and the query doesn't explicitly specify hidden columns to be returned.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
Very impressive! So sorry it's taken me so long to review.

Have you confirmed the performance gains with microbenchmarks?

As I mentioned offline, I have somewhat low confidence in my review since I'm lacking so much context. I just highlighted a couple places where you might help future readers gain some of that context with some high level comments.

> Reviewed 3 of 21 files at r1, 2 of 17 files at r14, 4 of 20 files at r19, 15 of 15 files at r23, 13 of 13 files at r24, 1 of 1 files at r25, 1 of 1 files at r26, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MmJhR_147A_tq7N4SKI:b-iidpjm)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 101 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJACORDxpWns_xE4cj:-MmJACORDxpWns_xE4ck:b-8gu56d) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner.go#L101)):*
> ```Go
> 	}
> 	var willEmit int
> 	if c.needLeftTuples {
> ```

all of this logic below seems a bit complex and low level. Is there a way to abstract it a bit? For example, could we make a helper function called `numTuplesToEmit`, and if it returns 0 then call `emitFirstZeroBatch`?
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 349 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJDXcw8XVjv8ZZOr18:-MmJDXcw8XVjv8ZZOr19:bpf05qk) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner.go#L349)):*
> ```Go
> 		left, right cjMutableBuilderState
> 
> 		// numAlreadyEmitted tracks the number of joined rows returned based on
> ```

nit: I feel like these variable names are slightly confusing. Would something like `numEmittedCurLeftBatch` and `numEmittedTotal` accurately describe what these mean?
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 165 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJFyy3BgDhhWDYBx9C:-MmJFyy3BgDhhWDYBx9D:b-kg66qu) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L165)):*
> ```Go
> //
> // The goal of this method is to repeat each tuple from the left input
> // leftNumRepeats times. Only the tuples in [curSrcStartIdx; leftSrcEndIdx) are
> ```

nit: `;` -> `,`
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 177 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJGAVi7mk_w8hlSXC7:-MmJGAVi7mk_w8hlSXC8:b-9ynwhh) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L177)):*
> ```Go
> 				buildFromLeftBatch(b, currentBatch, sel, true)
> 			} else {
> 				buildFromLeftBatch(b, currentBatch, sel, false)
> ```

nit: might be clearer to pass `nil` instead of `sel`.

nit: `false /* hasSel */` (ditto for above)
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 93 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJILVR6DjwFLMyqEKy:-MmJILVR6DjwFLMyqEKz:b-ftik05) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/mergejoiner.go#L93)):*
> ```Go
> 	leftBatchDone bool
> 	// rightFirstTuple is the first tuple of the right buffered group. It is set
> 	//only in case the right buffered group spans more than one input batch.
> ```

nit: add a space before only
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 120 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJmsYM1DMhjn5IimU9:-MmJmsYNDgMiXNWsHCHo:bi1rll1) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L120)):*
> ```Go
> type mjState int
> 
> const (
> ```

I feel that this might be a good place to add a comment that gives a bit of a higher level picture of what's happening in the merge joiner, by describing what state transitions are possible, and what each phase does at a very high level. I think there is also some vocabulary that would be helpful to define, such as what is the probing state / batches, what's a batch v. a buffered group, etc. 
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 642 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJjWLqDzZQfryVuNz7:-MmJjWLqDzZQfryVuNz8:bwgx9m4) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L642)):*
> ```Go
> 	if needToModify {
> 		colexecutils.UpdateBatchState(
> 			rBatch, rLength, rBatchHasSel, o.bufferedGroup.scratchSel,
> ```

Is this correct in both cases  where `rBatchHasSel` is true or false? Doesn't seem like `scratchSel` is always set.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 682 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJvB-Z0D1gj1NSdqkH:-MmJvB-a1fsog5pY6yr4:bm4g7rb) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L682)):*
> ```Go
> 	// It is ok that we might call Init() multiple times - it'll be a noop after
> 	// the first one.
> 	o.left.distincter.Init(o.Ctx)
> ```

can you give some intuition about why we need a distincter
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 640 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJMcwc08BIWDGFFAwx:-MmJMcwd3FY1A7FRQMZ0:b-75ywm7) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L640)):*
> ```Go
> func (o *mergeJoin_JOIN_TYPE_STRINGOp) probeBodyLSel_IS_L_SELRSel_IS_R_SEL() {
> 	lSel := o.proberState.lBatch.Selection()
> 	_ = lSel
> ```

why is this needed?
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 1130 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJwNhR0KmD0UGzs5_H:-MmJwNhSALeHySTWYa9E:bkvlqqt) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L1130)):*
> ```Go
> }
> 
> // transitionIntoBuildingFromBufferedGroup should be called once we have
> ```

I feel like I've read over the two functions that follow a couple of times and still feel like I'm missing something. Maybe the high level comment I suggested in another place giving more context on the phase transitions and vocabulary will help, but consider whether there is something else you can add here to help readers understand what follows.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
Thanks for taking a look!

> Have you confirmed the performance gains with microbenchmarks?

Yeah, the benchmarks are [here](https://github.com/cockroachdb/cockroach/pull/68921#issuecomment-899992435).

I haven't responded to the comments about adding more explanations in the code yet, but I will.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MmKU2VN3EsH2HRUN6Cp:b51v21r)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 101 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJACORDxpWns_xE4cj:-MmKXKBjEczkdmOJyjG5:bb1fpe0) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner.go#L101)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

all of this logic below seems a bit complex and low level. Is there a way to abstract it a bit? For example, could we make a helper function called `numTuplesToEmit`, and if it returns 0 then call `emitFirstZeroBatch`?
</blockquote></details>

I extracted a helper method, but I don't know how to simplify the logic inside of it.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 349 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJDXcw8XVjv8ZZOr18:-MmKXldh3qWMk2Zup48u:b5xqel8) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/crossjoiner.go#L349)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

nit: I feel like these variable names are slightly confusing. Would something like `numEmittedCurLeftBatch` and `numEmittedTotal` accurately describe what these mean?
</blockquote></details>

Yeah, that sounds right, I like it, done.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 642 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJjWLqDzZQfryVuNz7:-MmKa5Wc9a-UkrXDIAPM:b-b6wl35) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L642)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is this correct in both cases  where `rBatchHasSel` is true or false? Doesn't seem like `scratchSel` is always set.
</blockquote></details>

Yes, if `rBatchHasSel` is `false`, then `scratchSel` is ignored.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 682 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJvB-Z0D1gj1NSdqkH:-MmKZi9rDcpibj9g6y73:bqxjn6v) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L682)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

can you give some intuition about why we need a distincter
</blockquote></details>

Imagine that the first batch from the left was `1, 1, 1, 2`. We process first three `1`s directly from the batch and put the last `2` into the buffered group. Then we build the cross product based on a single row `2` from the left and the whole buffered group on the right.

Now we have to fetch the next batch from the left, suppose it is `2, 2, 2, 2` - all of the rows are still part of the same buffered group, so we populate the cross product of these 4 rows against the whole right buffered group. Next we fetch another batch from the left, suppose it is `2, 3, 3, 3` - here only the first row is still part of the same buffered group.

We use the distincter (which is an ordered distinct essentially) to find the first row in the current left batch that is different from the current left buffered group. In `isBufferedGroupFinished` we check whether the 0th row from the batch is different, and if it's not, then we run the `distincter` on the whole batch. If the distincter doesn't find any new distinct rows, then all rows from the batch are the same on the equality columns as the very first row that started the left buffered group, so the group continues. Once we find an actual distinct row, the left buffered group's boundary has been found.

Hope this makes sense.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 640 at r23](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJMcwc08BIWDGFFAwx:-MmKV1L67bxpSZwt_8aN:b-91l02p) ([raw file](https://github.com/cockroachdb/cockroach/blob/dfd31e3ea716ca649b087be9a4d7c3b1b4aad763/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L640)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

why is this needed?
</blockquote></details>

At some point, this variable was unused in some branches (we're in a template that is instantiated 4 times), but this is no longer the case - we now always use both `lSel` and `rSel` in all branches. Removed.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 120 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJmsYM1DMhjn5IimU9:-MmPscvFAzx4i31donRe:b-rdrf5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L120)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I feel that this might be a good place to add a comment that gives a bit of a higher level picture of what's happening in the merge joiner, by describing what state transitions are possible, and what each phase does at a very high level. I think there is also some vocabulary that would be helpful to define, such as what is the probing state / batches, what's a batch v. a buffered group, etc. 
</blockquote></details>

Alright, I've added a pretty extensive comment here which attempts to define some terminology as well as includes an example that describes the flow of the algorithm. Please let me know what you think and whether there something that I missed.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 1130 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJwNhR0KmD0UGzs5_H:-MmPtuRm7GUnmmbS61CE:bkr6zro) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L1130)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

I feel like I've read over the two functions that follow a couple of times and still feel like I'm missing something. Maybe the high level comment I suggested in another place giving more context on the phase transitions and vocabulary will help, but consider whether there is something else you can add here to help readers understand what follows.
</blockquote></details>

I hope that large comment does provide the meaningful context in order to understand what these methods are about. I'm not sure what I can say more / differently here though.

One thing that definitely helps with reading of this code is thinking only about the INNER join (at least initially). With that assumption in mind, most code in these two methods is not applicable (i.e. it is templated away), so it's easier to see the actual meaning of the methods in the common cases.


<!-- Sent from Reviewable.io -->

rytaft reviewed (commented):
> Reviewed 3 of 17 files at r27, 17 of 17 files at r31, 13 of 13 files at r32, 4 of 4 files at r34, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 199 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTG1HP1V7I63QJxOyz:-MmTG1HQ471a5jT2hVJl:byue15g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L199)):*
> ```Go
> 	case descpb.RightOuterJoin:
> 		leftBatch := c.inputOne.Next()
> 		c.builderState.left.currentBatch = leftBatch
> ```

Isn't this assignment already happening inside `prepareForNextLeftBatch`?
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 204 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTFtjE2Lb2RQ11uENF:-MmTFtjF6uGxTnTwgVUm:byue15g) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L204)):*
> ```Go
> 	case descpb.FullOuterJoin:
> 		leftBatch := c.inputOne.Next()
> 		c.builderState.left.currentBatch = leftBatch
> ```

Isn't this assignment already happening inside `prepareForNextLeftBatch`?
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 220 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTHj5N2hmMNJ1sIrRY:-MmTEZz85eJ6g2nQjg-F:b-86a72p) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L220)):*
> ```Go
> 				// The current left batch is the first one that contains tuples
> 				// without a "match".
> 				c.prepareForNextLeftBatch(leftBatch, c.numRightTuples, leftBatch.Length())
> ```

Don't you always want to call this function after calling `c.inputOne.Next()`? (if the startIdx, `c.numRightTuples` in this case, is greater than the `endIdx`, you could potentially return early from that function.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 397 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTEjkP66iQjWsX01vi:-MmTEjkP66iQjWsX01vj:b1ngnij) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L397)):*
> ```Go
> // batch coming from the left input. Only rows with ordinals in
> // [startIdx, endIdx) range will be used for the cross join.
> func (b *crossJoinerBase) prepareForNextLeftBatch(batch coldata.Batch, startIdx, endIdx int) {
> ```

It seems like you could simplify some logic in other places if you have this function call `c.inputOne.Next()`. Then you only need to pass in the `startIdx`, since it seems like you never call this function with `endIdx` equal to anything other than `batch.Length()`.

If it's helpful, you could have this function return the batch length.

Edit: If this isn't possible (e.g., seems like you might need to call this function from the merge joiner?), maybe you should just create a wrapper around this function that does some of what I'm suggesting.
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 120 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJmsYM1DMhjn5IimU9:-MmT9XZFB8sW3Il2tEtD:bhpiorf) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L120)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Alright, I've added a pretty extensive comment here which attempts to define some terminology as well as includes an example that describes the flow of the algorithm. Please let me know what you think and whether there something that I missed.
</blockquote></details>

So helpful!! Thank you!
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 682 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJvB-Z0D1gj1NSdqkH:-MmT9xJi-hqIDqydzN16:b-989yli) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner.go#L682)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Imagine that the first batch from the left was `1, 1, 1, 2`. We process first three `1`s directly from the batch and put the last `2` into the buffered group. Then we build the cross product based on a single row `2` from the left and the whole buffered group on the right.

Now we have to fetch the next batch from the left, suppose it is `2, 2, 2, 2` - all of the rows are still part of the same buffered group, so we populate the cross product of these 4 rows against the whole right buffered group. Next we fetch another batch from the left, suppose it is `2, 3, 3, 3` - here only the first row is still part of the same buffered group.

We use the distincter (which is an ordered distinct essentially) to find the first row in the current left batch that is different from the current left buffered group. In `isBufferedGroupFinished` we check whether the 0th row from the batch is different, and if it's not, then we run the `distincter` on the whole batch. If the distincter doesn't find any new distinct rows, then all rows from the batch are the same on the equality columns as the very first row that started the left buffered group, so the group continues. Once we find an actual distinct row, the left buffered group's boundary has been found.

Hope this makes sense.
</blockquote></details>

Yep! Thanks!
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 150 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmT92eM2loGM0kv-5md:-MmT92eN28sVehNYlpm4:b1z8o7a) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e6125e95ae934b837c82aebf38570333da8aa35/pkg/sql/colexec/colexecjoin/mergejoiner.go#L150)):*
> ```Go
> // right buffered group.
> //
> // At this point, we have exhausted both inputs, and we're done.
> ```

Thank you for this comment!! So much clearer now! 👏👏👏
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 1130 at r26](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmJwNhR0KmD0UGzs5_H:-MmTNtTpASAgB1jISwxf:b-fqm331) ([raw file](https://github.com/cockroachdb/cockroach/blob/75dfd993d07949bf79287c350fe9c44bc62b766a/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L1130)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

I hope that large comment does provide the meaningful context in order to understand what these methods are about. I'm not sure what I can say more / differently here though.

One thing that definitely helps with reading of this code is thinking only about the INNER join (at least initially). With that assumption in mind, most code in these two methods is not applicable (i.e. it is templated away), so it's easier to see the actual meaning of the methods in the common cases.
</blockquote></details>

Yep, this helps. Thanks!
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 1130 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTAAOSDSa_oUosruJ1:-MmTAAOTEOjpwe5zEjYm:bem97d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e6125e95ae934b837c82aebf38570333da8aa35/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L1130)):*
> ```Go
> // exhausted).
> //
> // For a more detail explanation and an example please refer to the comment at
> ```

nit: detail -> detailed
___
*[pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go, line 1191 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTAF162yJFe7ogpEvy:-MmTAF162yJFe7ogpEvz:bem97d6) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e6125e95ae934b837c82aebf38570333da8aa35/pkg/sql/colexec/colexecjoin/mergejoiner_tmpl.go#L1191)):*
> ```Go
> // It is assumed that transitionIntoBuildingFromBufferedGroup has been called.
> //
> // For a more detail explanation and an example please refer to the comment at
> ```

nit: detail -> detailed


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 199 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTG1HP1V7I63QJxOyz:-MmTyiqtDctPw8t_i5Uz:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L199)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Isn't this assignment already happening inside `prepareForNextLeftBatch`?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 204 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTFtjE2Lb2RQ11uENF:-MmTyj8B62IO22lfZuLy:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L204)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Isn't this assignment already happening inside `prepareForNextLeftBatch`?
</blockquote></details>

Done.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 220 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTHj5N2hmMNJ1sIrRY:-MmTzQZ74Be0kPTz3A2-:b-lsfsqn) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L220)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Don't you always want to call this function after calling `c.inputOne.Next()`? (if the startIdx, `c.numRightTuples` in this case, is greater than the `endIdx`, you could potentially return early from that function.
</blockquote></details>

Hm, I think it is possible to refactor the code so that we could call `prepareForNextLeftBatch` with `startIdx > endIdx`. However, I think it'll be more confusing overall.

My thinking for the current code is as follows: for EXCEPT ALL we perform the "subtraction" of the right rows from the left rows by skipping first `len(right rows)` from the left side. We do so explicitly here in `setupForBuilding`, and for all these "subtracted" rows we don't have any output to emit, so there is no point in calling `prepareForNextLeftBatch` on these "subtracted" batches.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 397 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTEjkP66iQjWsX01vi:-MmU0lGR4KnoXLMhBxRS:btc3ozc) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L397)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

It seems like you could simplify some logic in other places if you have this function call `c.inputOne.Next()`. Then you only need to pass in the `startIdx`, since it seems like you never call this function with `endIdx` equal to anything other than `batch.Length()`.

If it's helpful, you could have this function return the batch length.

Edit: If this isn't possible (e.g., seems like you might need to call this function from the merge joiner?), maybe you should just create a wrapper around this function that does some of what I'm suggesting.

</blockquote></details>

Yeah, we cannot do that for the merge joiner, but I like your suggestion, so I added a helper that is now used in most places, thanks!
___
*[pkg/sql/colexec/colexecjoin/mergejoiner.go, line 150 at r31](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmT92eM2loGM0kv-5md:-MmU1Rva-2Q900IVrk1U:bla7224) ([raw file](https://github.com/cockroachdb/cockroach/blob/8e6125e95ae934b837c82aebf38570333da8aa35/pkg/sql/colexec/colexecjoin/mergejoiner.go#L150)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Thank you for this comment!! So much clearer now! 👏👏👏
</blockquote></details>

:smiley: 


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> great work! (But maybe wait for @michae2's approval too)

> Reviewed 18 of 18 files at r35, 13 of 13 files at r36, 4 of 4 files at r38, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MmUO_9V4oxvSbSR2dh4:b-qo6280)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 220 at r34](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmTHj5N2hmMNJ1sIrRY:-MmUNstC86I-Fe8xwnPX:bqqoem6) ([raw file](https://github.com/cockroachdb/cockroach/blob/6dd3d22183b4a4add47dcaa631532a364faf5816/pkg/sql/colexec/colexecjoin/crossjoiner.go#L220)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, I think it is possible to refactor the code so that we could call `prepareForNextLeftBatch` with `startIdx > endIdx`. However, I think it'll be more confusing overall.

My thinking for the current code is as follows: for EXCEPT ALL we perform the "subtraction" of the right rows from the left rows by skipping first `len(right rows)` from the left side. We do so explicitly here in `setupForBuilding`, and for all these "subtracted" rows we don't have any output to emit, so there is no point in calling `prepareForNextLeftBatch` on these "subtracted" batches.
</blockquote></details>

Ok, makes sense


<!-- Sent from Reviewable.io -->

michae2 reviewed (changes requested):
This gives me a headache, but it's also exciting. :slightly_smiling_face: One day we could go even further into headache land and turn this into [block nested loop join](https://en.wikipedia.org/wiki/Block_nested_loop) for fewer iterations of the right side.

Still reading, but I had a few questions. Sorry I'm so slow.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MmV3zjF4HrNSPXvEJCW:b-gc6uhz)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 118 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUzxvD7nfStb7vXAiC:-MmUzxvD7nfStb7vXAiD:b-jfe2uo) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner.go#L118)):*
> ```
> c.buildFromLeftInput(c.Ctx, 0 /* destStartIdx */)
> ```

Do I understand it correctly that `buildFromLeftInput` might not fill the entire output batch (if it reaches the end of the current left batch), but `buildFromRightInput` will always fill the entire output batch (unless it reaches the end of the join results)?

It seems like we account for this possibility with `c.output.SetLength(willEmit)` below, but by the time we do that, `builderState.right` has already been updated as if the entire output batch was filled.

Seems like we need to pass `willEmit` into `buildFromRightInput`? Or maybe more simply call `c.output.SetLength(willEmit)` before calling `buildFromRightInput`?

Hopefully I'm just misunderstanding.
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 207 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUtejICrk3QmVT6zWM:-MmUtejICrk3QmVT6zWN:b-wnf7xg) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L207)):*
> ```Go
> 				currentBatch := bs.right.currentBatch
> 				if currentBatch == nil {
> 					currentBatch, err = b.rightTuples.Dequeue(ctx)
> ```

Question: Do we not have to worry about `currentBatch.Selection()` on the right side because `SpillingQueue.Enqueue()` only keeps selected rows?


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> One day we could go even further into headache land and turn this into block nested loop join for fewer iterations of the right side.

Oh, I'm already scared of that complexity lol.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MmVNL-i0MarEDf0QBHq:b4llh7e)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 118 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUzxvD7nfStb7vXAiC:-MmVO23F-61omhjzpFLX:bvr2z6l) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner.go#L118)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> ```
> c.buildFromLeftInput(c.Ctx, 0 /* destStartIdx */)
> ```

Do I understand it correctly that `buildFromLeftInput` might not fill the entire output batch (if it reaches the end of the current left batch), but `buildFromRightInput` will always fill the entire output batch (unless it reaches the end of the join results)?

It seems like we account for this possibility with `c.output.SetLength(willEmit)` below, but by the time we do that, `builderState.right` has already been updated as if the entire output batch was filled.

Seems like we need to pass `willEmit` into `buildFromRightInput`? Or maybe more simply call `c.output.SetLength(willEmit)` before calling `buildFromRightInput`?

Hopefully I'm just misunderstanding.
</blockquote></details>

There are two exit conditions for both of the `buildFrom*Input` methods:
- we are always making sure that we never exceed `c.output.Capacity()`
- and we specify the precise values of `setup.leftNumRepeats` and `setup.rightNumRepeats`.

I think in the example scenario you mention, when the cross product with the current left batch doesn't fill the output up to capacity, `rightNumRepeats` will be set in such a manner that `bs.right.numRepeatsIdx < bs.setup.rightNumRepeats` will be `false`, so we'll exit out of `buildFromRightInput` before filling up to the capacity.

The `rightNumRepeats` is setup in `prepareForNextLeftBatch` (it depends on each left batch) whereas `leftNumRepeats` is set in `setupLeftBuilder` (it depends on the whole right input).
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 207 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUtejICrk3QmVT6zWM:-MmVNR8w3ev6wY5Vugu3:b-ix6rhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L207)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

Question: Do we not have to worry about `currentBatch.Selection()` on the right side because `SpillingQueue.Enqueue()` only keeps selected rows?
</blockquote></details>

Yes, exactly - that's the contract of the spilling queue. Do you think it's worth adding a comment about that?


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
> Reviewed 1 of 21 files at r1, 2 of 20 files at r19, 2 of 17 files at r27, 3 of 18 files at r35.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2, @rytaft, and @yuzefovich)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner.go, line 118 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUzxvD7nfStb7vXAiC:-Mm_5pZv9PcgfPFozCIN:b-9755fn) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner.go#L118)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

There are two exit conditions for both of the `buildFrom*Input` methods:
- we are always making sure that we never exceed `c.output.Capacity()`
- and we specify the precise values of `setup.leftNumRepeats` and `setup.rightNumRepeats`.

I think in the example scenario you mention, when the cross product with the current left batch doesn't fill the output up to capacity, `rightNumRepeats` will be set in such a manner that `bs.right.numRepeatsIdx < bs.setup.rightNumRepeats` will be `false`, so we'll exit out of `buildFromRightInput` before filling up to the capacity.

The `rightNumRepeats` is setup in `prepareForNextLeftBatch` (it depends on each left batch) whereas `leftNumRepeats` is set in `setupLeftBuilder` (it depends on the whole right input).
</blockquote></details>

Thank you!
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 158 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmZxHfqB64Fm4Eb1pJt:-MmZxHfqB64Fm4Eb1pJu:b-nrgtht) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L158)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 		if colIdx < len(b.left.types)-1 {
> 			// Transition to building the next column with the same
> 			// initial builder state as the current column.
> 
> 
> 
> 			*bs = initialBuilderState
> 		}
> ```
</details>


nit: Instead of `bs := &b.builderState.left` at the top, could we use `bs := initialBuilderState` in the same place that we create `outStartIdx` so that it's clear we're operating on a copy for each column? (Then we would have to do the assignment back to `b.builderState.left` at the end.)
___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 207 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmUtejICrk3QmVT6zWM:-MmZyT-n28zoOkRo5CjH:b-es4ggg) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L207)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Yes, exactly - that's the contract of the spilling queue. Do you think it's worth adding a comment about that?
</blockquote></details>

No, it's ok, just asking.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 158 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmZxHfqB64Fm4Eb1pJt:-MmdWQLWCHMQnI3ASqqT:b8dnzvo) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L158)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> ```
> 		if colIdx < len(b.left.types)-1 {
> 			// Transition to building the next column with the same
> 			// initial builder state as the current column.
> 
> 
> 
> 			*bs = initialBuilderState
> 		}
> ```

nit: Instead of `bs := &b.builderState.left` at the top, could we use `bs := initialBuilderState` in the same place that we create `outStartIdx` so that it's clear we're operating on a copy for each column? (Then we would have to do the assignment back to `b.builderState.left` at the end.)
</blockquote></details>

Hm, yeah, I think it'd be clearer this way, thanks. I also refactored the code a bit more, and it looks much simpler now.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)

___
*[pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go, line 158 at r35](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-MmZxHfqB64Fm4Eb1pJt:-Mmdt-Ft7vvXUejgHXul:b-36om85) ([raw file](https://github.com/cockroachdb/cockroach/blob/42a3e9f6bc186cb7c669c58fd9c75e1caf97e4f8/pkg/sql/colexec/colexecjoin/crossjoiner_tmpl.go#L158)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

Hm, yeah, I think it'd be clearer this way, thanks. I also refactored the code a bit more, and it looks much simpler now.
</blockquote></details>

Man, these copies on each column show up noticeably in the microbenchmarks:
```
name                                                       old alloc/op   new alloc/op     delta
CrossJoiner/spillForced=false/type=INNER/rows=1-24           93.0kB ± 0%      91.1kB ± 0%      -2.08%  (p=0.000 n=10+10)
CrossJoiner/spillForced=false/type=INNER/rows=16-24           113kB ± 1%       110kB ± 0%      -2.39%  (p=0.000 n=10+9)
CrossJoiner/spillForced=false/type=INNER/rows=256-24          188kB ± 0%       184kB ± 0%      -2.16%  (p=0.000 n=10+10)
CrossJoiner/spillForced=false/type=INNER/rows=2048-24         339kB ± 0%       778kB ± 0%    +129.36%  (p=0.000 n=10+9)
CrossJoiner/spillForced=false/type=INNER/rows=8192-24         863kB ± 0%      8915kB ± 0%    +933.35%  (p=0.000 n=9+10)
```
```
ROUTINE ======================== github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin.(*crossJoinerBase).buildFromLeftInput.func1 in /Users/yuzefovich/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/colexecjoin/crossjoiner.eg.go
  760.52MB   760.52MB (flat, cum) 93.27% of Total
         .          .    692:					var srcStartIdx int
         .          .    693:					// Loop over every column.
         .          .    694:					for colIdx := range b.left.types {
         .          .    695:						// Get a fresh copy of the builder state so that each column has a fresh
         .          .    696:						// start.
  760.52MB   760.52MB    697:						bs := b.builderState.left
         .          .    698:						if colIdx == len(b.left.types)-1 {
         .          .    699:							// This is the last column. We need to update the builder state so
         .          .    700:							// that we continue where we left off on the next call.
         .          .    701:							defer func() {
         .          .    702:								b.builderState.left = bs
```
So I partially reverted the change. However, overall I think it has gotten cleaner.


<!-- Sent from Reviewable.io -->

michae2 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> nice work!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68921#-:-MnD_ejz4-YjRgEOzSOL:b-kbynj8)* status: :shipit: complete! 1 of 0 LGTMs obtained (and 1 stale) (waiting on @michae2 and @rytaft)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
Thanks for the reviews!

bors r+

yuzefovich commented:
A flake.

bors r+

yuzefovich commented:
Same flake.

bors r+
